File: code-salon/src/salon.py
import cpp_stylist as cpps
import argparse
import sys
import re
import os

parser = argparse.ArgumentParser(description='Style the source code.')
parser.add_argument('input_file', help='the input source code')
parser.add_argument('output_file', help='the output source code')
parser.add_argument('--lang', help='the language of the source code', default='cpp')
parser.add_argument('--new-line', help='if you want new line before parentheses, default is false', default=False)
parser.add_argument('--delete-empty-lines', help='if you want to delete the empty lines, default is false', default=False)
args = parser.parse_args()

non_escaped_double_quotes = re.compile(r"(?<!\\)\"")
non_line_tail_semi_colons = re.compile(";[^\\n]")
multi_line_comment = re.compile("\\/\\*|\\*\\/")
single_line_comment = re.compile(r"//")
#string_regex = re.compile(r'/"(?:[^"\\]|\\.)*"/')
string_regex = re.compile(r'.+')

def str_to_bool(s):
	if(not s):
		return False;
	s = s.lower()
	if(s == 'true' or s == '1'):
		return True
	else:
		return False # default is false
		
delete_empty_lines = str_to_bool(args.delete_empty_lines)

def construct_option_string():
	s = ""
	if(str_to_bool(args.new_line)):
		s += 'new_line=' + '1'
	else:
		s += 'new_line=' + '0'
	return s
	
def toggle(flag):
	return not flag
	
def add_new_lines_after_semi_colons(s):
	result = ""
	lines = non_line_tail_semi_colons.split(s)
	for line in lines:
		result += line + ';' + '\n'
	return result
	
def process_line(line):
	result = ""
	words = line.split()
	if(len(words) > 0):
		result += words[0]
		for word in words[1:]:
			if(word != '{' and word != '}'):
				result += ' ' + word
			else:
				result += word
	if(line.endswith(';')):
		result += '\n'
	return result
	
def process_non_string(s):
	result = ""
	lines = s.split('\n')
	for line in lines:
		if(line.startswith('#') or single_line_comment.search(line)): # lines starts with '#', like #include ..., we just copy it
			result += line.strip() + '\n'
		elif(line.strip() == ''): # empty line
			result += '\n'
		else:
			result += process_line(line)
	return result
	
# process the source code line-by-line to clean the code
def get_clean_string(source):
	result = ""
	in_comment = False
	
	comment_slices = multi_line_comment.split(source)
	for comment_slice in comment_slices:
		if(not in_comment):
			in_string = False
			string_slices = non_escaped_double_quotes.split(comment_slice)
			
			for slice in string_slices:
				if(not in_string):
					result += process_non_string(slice)
					in_string = not in_string
				else:
					result += ' "'
					result += slice
					result += '" '
					in_string = not in_string
			in_comment = not in_comment
		else:
			result += '/*'
			result += comment_slice
			result += '*/'
			in_comment = not in_comment
			
	return result

def main():
	in_file = open(args.input_file)
	
	if os.path.dirname(args.output_file) != '':
		os.makedirs(os.path.dirname(args.output_file), exist_ok=True)
	out_file = open(args.output_file, 'w+')
	
	stylist = cpps.CppStylist(construct_option_string())
	
	source = in_file.read()
	cleaned_source = get_clean_string(source)
	#print(cleaned_source)
	styled = stylist.style(cleaned_source)
	
	out_file.write(styled)
	
	
	

if __name__ == "__main__":
	main()

File: code-salon/src/salon.go
package main

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"flag"
	"fmt"
)

const python string = "python3"
var script string
var pwd string

func salonFile(inFile string, outFile string, options []string) {
	py, _ := exec.LookPath(python)
	var args []string = []string {py, script}
	args = append(args, options...)
	args = append(args, inFile)
	args = append(args, outFile)
	pyCmd := &exec.Cmd {
		Path: py,
		Args: args,
		Stdout: os.Stdout,
		Stderr: os.Stdin,
	}
	pyCmd.Run()
}

func salonFileConcurrent(inFile string, outFile string, options []string, wg *sync.WaitGroup) {
	defer wg.Done()
	salonFile(inFile, outFile, options)
}

func singleMode(inFile string, outFile string, options []string, lang string) {
	if !strings.HasSuffix(inFile, "." + lang) {
		fmt.Println("WARNING: " + inFile + " is not a ." + lang + " file.")
	}
	salonFile(inFile, outFile, options)
}

func concurrentMode(inDir string, outDir string, options []string, lang string) {
	var wg sync.WaitGroup
	
	var inputFiles []string
	filepath.Walk(inDir, func(path string, info os.FileInfo, err error) error {
		inputFiles = append(inputFiles, path)
		return nil
	})
	inputFiles = inputFiles[1:]
	
	var filenames []string
	for _, file := range inputFiles {
		tmp := strings.Split(file, "/")
		filename := tmp[len(tmp) - 1]
		filenames = append(filenames, filename)
	}
	
	var outputFiles []string
	for _, filename := range filenames {
		outputFile := outDir + "/" + filename
		outputFiles = append(outputFiles, outputFile)
	}
	
	
	for i, inFile := range inputFiles {
		if strings.HasSuffix(inFile, "." + lang) {
			wg.Add(1)
			go salonFileConcurrent(inFile, outputFiles[i], options, &wg)
		} else {
			fmt.Println(inFile + " is not a ." + lang + " file, ignoring.")
		}
	}
	wg.Wait()
}

func main() {
	ex, _  := os.Executable()
	pwd := filepath.Dir(ex)
	script = pwd + "/code-salon-deps/salon.py"
	
	newLine := flag.String("new-line", "0", "if you want a new line before parentheses")
	deleteEmptyLines := flag.String("delete-empty-lines", "0", "if you want to delete empty lines")
	indent := flag.String("indent", "0", "number of spaces for an indent. If this is 0, then a tab is used.")
	input := flag.String("in", "", "the input file/directory")
	output := flag.String("out", "", "the output file/directory")
	lang := flag.String("lang", "cpp", "the programming language of the source file, default is cpp")
	flag.Parse()
	in := string(*input)
	out := string(*output)
	
	if in == "" || out == "" {
		fmt.Fprintf(os.Stderr, "Usage: %s -in=<input_file> -out=<output_file>\n", os.Args[0])
		flag.PrintDefaults()
		os.Exit(1)
	}
	
	var options []string
	options = append(options, []string{"--new-line", string(*newLine)}...)
	options = append(options, []string{"--delete-empty-lines", string(*deleteEmptyLines)}...)
	options = append(options, []string{"--indent", string(*indent)}...)
	options = append(options, []string{"--lang", string(*lang)}...)
	
	fi, err := os.Stat(in)
	if err != nil {
		panic(err)
	}
	switch mode := fi.Mode(); {
	case mode.IsDir():
		concurrentMode(in, out, options, string(*lang))
	case mode.IsRegular():
		singleMode(in, out, options, string(*lang))
	}
}